### DISCUSSION TITLE
FreeBSD port

### DISCUSSION THREAD

**Discussion by cyring** [cyring/CoreFreq#249] [Ideas]
# Purpose
Based on https://github.com/cyring/CoreFreq/commit/7e9a355e9b5429fb992b36d6348ceb606ae2e472#r51613397 the followings are about the attempt to port _CoreFreq_ to FreeBSD
# Impacts
## Compiler
Starting with FreeBSD 10.X clang is installed as cc
[Compiling with cc](https://docs.freebsd.org/en/books/developers-handbook/tools/#tools-compiling)
* As of today, _CoreFreq_ Daemon and Client can be built with `clang` version 11.1 for Linux

```
make CC='clang' clean corefreqd corefreq-cli
```
* Unfortunately the kernel module build fails with those errors:

```
make[1]: Entering directory '/usr/lib/modules/5.12.8-arch1-1/build'
  CC [M]  src/CoreFreq/corefreqk.o
clang-11: error: unknown argument: '-fplugin-arg-structleak_plugin-byref-all'
clang-11: error: unsupported option '-mrecord-mcount' for target 'x86_64-pc-linux-gnu'
```

**Comment by danfe** [comment]
User-land utilities currently require some patching, but not too much; apart from avoiding peeking into opaque `locale_t` type, there are things like different order of the `qsort_r(3)` arguments, some missing Linux-specific signals, non-POSIX `strerror_r(3)` in Glibc, etc.  I can upload them somewhere if there's interest, they allow those programs to build and work on FreeBSD, however, they are not very useful without the bigger challenge which is the kernel module itself, and I've just started this work.

**Reply by cyring** [reply]
I also have a few FreeBSD cases from previous project XFreq

* [I/O calls](https://github.com/cyring/xfreq/blob/4a95bc4082adeb11bdd87e60a0a2bc0f1bee1a43/svr/xfreq-intel.c#L1101)
* Basic [pthreads](https://github.com/cyring/xfreq/blob/4a95bc4082adeb11bdd87e60a0a2bc0f1bee1a43/cli/xfreq-cli.h#L10)
